<a href="https://colab.research.google.com/github/LuckyStrike42/Analitica/blob/main/Reto_2_%C3%81rboles_de_Decisi%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Árbol de desición Reto 2**

De acuerdo con el enunciado del Ejercicio 36 (Travel_Plan) de la Sección de Ejercicios 13.4 del Libro Guía Analítica de Negocios – Comunicación con Datos (Jaggia – Mc Graw Hill), Construir un modelo que permita predecir si un cliente utilizará la tarjeta de crédito para el pago de los gastos de viaje y de los alimentos

# Librerias y Datos


0. Cargar las librerias necesarias para la ejecución del modelo.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt # Graficador
from sklearn.tree import DecisionTreeClassifier # Arbol de desicion.
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import seaborn as sns

# Para no tener que cargar el archivo
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Carga de los datos

In [ ]:
nxl = '/content/drive/MyDrive/Analisis de negocios int/2. BD1_Travel_Data_int.xlsx'
df = pd.read_excel(nxl,sheet_name=0)
df = df.dropna()
df.head(100)

# Convertir la data de las columnas CedirCard & College en 1 y 0, Yes = 1 & No = 0
df['College'] = df['College'].map({'Yes':1,'No':0})
df['CreditCard'] = df['CreditCard'].map({'Yes':1,'No':0})
#Seleccion de las variables

dfx = np.array (df[['College','FoodSpend','Income','TravelSpend']]) #np.array = convertir el texto en numeros.
dfy = np.array(df['CreditCard'])


# Creacion del modelo.

2. árbol de desición.

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4) # 4 variables de entrada (4 NIVELES)
mar.fit(dfx,dfy)
dfy_pred=mar.predict(dfx)

from sklearn.metrics import confusion_matrix
cm = confusion_matrix(dfy,dfy_pred)
print(cm)

#    0   1
# 0  VN  FP
# 1  FN  VP

VN = cm[0,0]
FP = cm[0,1]
FN = cm[1,0]
VP = cm[1,1]

exactitud=(VP+VN)/(VP+VN+FP+FN) #Comportamiento general
sensibilidad=VP/(VP+FN) #Como se comporto frente a los positivos
especificidad=VN/(VN+FP) #Como se comporto frente alos negativos


print('Exactitud:')
print(exactitud)
print('Sensibilidad:')
print(sensibilidad)
print('Especificidad:')
print(especificidad)

[[334   9]
 [ 42 115]]
Exactitud:
0.898
Sensibilidad:
0.732484076433121
Especificidad:
0.9737609329446064


# Árbol de desición.

3. Árbol PNG

In [ ]:
#Librerias para graficar arbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend','Income','TravelSpend'] #Variables de entrada o Titulos para el arbol
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')


True

# Nuevos clientes

4. Pronosticos de los nuevos clientes

In [ ]:
df2 = pd.read_excel(nxl,sheet_name=1)
df2 = df2.dropna()
df2.head(100)

#Convertir la data de la columna College en 1 y 0, Yes = 1 & No = 0
df2['College'] = df2['College'].map({'Yes':1,'No':0})
df2.head(100)

df2x=np.array(df2[['College','FoodSpend','Income','TravelSpend']]) #np.array = convertir el texto en numeros.
dfy2=mar.predict_proba(df2x)
print(dfy2)

#    0   1
# 0  VN  FP
# 1  FN  VP

VN = dfy2[0,0]
FP = dfy2[0,1]
FN = dfy2[1,0]
VP = dfy2[1,1]

#De acuerdo con los nuevos clientes en la base de datos Travel_Plan_Score, Cuál es la probabilidad de que el primer cliente tenga planes de viajar el próximo año, y Cuá es la probabilidad para el segundo cliente.
print('La probabilidad de que el primer cliente use su trajeta de credito para pagar sus gastos de viaje y de los alimentos en sus planes de viajar el proximo ańo es:',FP*100,'%')
print('La probabilidad de que el segundo cliente use su trajeta de credito para pagar sus gastos de viaje y de los alimentos en sus planes de viajar el proximo ańo es:',VP*100,'%')






[[1.         0.        ]
 [0.78676471 0.21323529]]
La probabilidad de que el primer cliente use su trajeta de credito para pagar sus gastos de viaje y de los alimentos en sus planes de viajar el proximo ańo es: 0.0 %
La probabilidad de que el segundo cliente use su trajeta de credito para pagar sus gastos de viaje y de los alimentos en sus planes de viajar el proximo ańo es: 21.323529411764707 %


# **Relgas del negocio y nodos puros**

**Las reglas del negocio para los nuevos clientes:**

Cliente #1: TravelSpend > 1567.875 AND Income > 59098.0.



Cliente #2: Income <= 59098.0 AND FoodSpend > 2711.875 AND TravelSpend <= 3008.25.



**Las condiciones de los nodos puros son:**

1.⁠ ⁠Nodo [1,0] = TravelSpend <=1188.0 AND FoodSpend > 4972.55.

2.⁠ ⁠Nodo [0,47] = TravelSpend <=1188.0 AND FoodSpend > 4990.73.

3.⁠ ⁠Nodo [108,0] = TravelSpend > 1567.875 AND Income > 59098.0


# **Análisis de Resultados**

El modelo tiene un rendimiento general sólido con una exactitud del 89.8%. Esto indica que clasifica correctamente la mayoría de los casos. Sin embargo, al desglosar el rendimiento, la sensibilidad (73.2%) muestra que el modelo tiene un rendimiento moderado para identificar correctamente a aquellos que poseen una tarjeta de crédito. Por otro lado, la especificidad es alta (97.4%), lo que significa que es muy preciso para identificar correctamente a aquellos que no poseen una tarjeta de crédito. Esto sugiere que el modelo tiende a ser más conservador en su clasificación, evitando falsos positivos.

En cuanto a los nuevos clientes, predice que el primer cliente tiene una probabilidad del 0% de utilizar su tarjeta de crédito para gastos de viaje, mientras que el segundo cliente tiene una probabilidad del 21.32%. Esto indica una baja intención del primer cliente de realizar gastos de viaje, posiblemente debido a su perfil financiero o hábitos de gasto. En cambio, el segundo cliente, con una mayor probabilidad de usar su tarjeta de crédito, podría estar más dispuesto a incurrir en estos gastos, lo cual coincide con las reglas de negocio establecidas en el modelo, donde factores como un mayor ingreso y gasto en alimentos influyen positivamente en la probabilidad de uso de la tarjeta para viajes.
